tensorflow/compiler/xla/service/gpu/gpu_conv_runner.cc

```diff
@@ -223,7 +223,16 @@ Status RunGpuConvImpl(const GpuConvParams& params,
   auto output_buf = se::DeviceMemory<OutputType>(params.output_buf);
   AlgorithmConfig algorithm = params.algorithm;
 
-  if (options.algo_override) {
+  // in ROCm mode, the first call to run the convolution needs to trigger the
+  // code that calls miopenFind* API. That triggger is implicit, it is based
+  // on whether or not the AlgorithmConfig::algorithm is empty! So for the
+  // first call we need to ensure that the AlgorithmConfig::algorithm is
+  // empty. For all subsequent calls, we should use the value retrieved from
+  // the backend_config
+  if ((options.algo_override.has_value()) &&
```

Comment:
Could we make this ROCm-specific? Or even better, could we have a cleaner workaround?

Reply:
Reverting this change makes CUDA tests pass again.

Reply:
As long as miopen find mode exist in `ROCm`, the passed in algorithm has to be empty for [this check to work](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/rocm/rocm_dnn.cc#L2623). If this is causing trouble from your end, I can add an additional condition in this if statement to check if the stream is under `se::PlatformKind::kROCm`.

Reply:
Could you expand on what is contained in the check (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/stream_executor/rocm/rocm_dnn.cc#L2623) and why is it necessary?

Reply:
This conditional check helps to determine if it is in profiling stage or not. In profiling stage, `ThenConvolveWithAlgorithm()` will call with [an empty `AlgorithmConfig()`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/conv_ops.cc#L1015). rocm_dnn use the empty AlgorithmConfig() to know it is in profiling stage (High overhead, for miopen to find and compile the algorithm). 

When profiling is over, AlgorithmConfig will not be empty and it includes both the algorithm and the scratch_size. This scenario has low overhead because all information are known so `miopen` can act on known parameters.

Reply:
Sorry I still don't quite understand. What is the difference between CUDA trace and ROCm trace? Why CUDA does not need the same trick?

Reply:
That's a fair question. If I remember correctly, this is due to `miopen` and `cudnn` library requires different arguments for getting a convolution algorithm. A detailed explanation please see [here](https://github.com/tensorflow/tensorflow/blob/10a42250f398a2d8b295a8d23603fc0286a67051/tensorflow/stream_executor/dnn.h). I was not involved in this part of code changes. @deven-amd if you need further information.

Reply:
Sorry I still do not understand the issue here. If there is a trigger required for ROCm, can we make it explicit and trigger it only in ROCm mode? Moreover, maybe we could trigger it inside the actual ROCm-specific function using e.g. `std::call_once`?

Reply:
I make it explicitly to be triggered under ROCm only. I believe this should address the failures caused by this.